design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,204 @@
+discussion.
+* Support for dates before Unix epoch - this is not supported by Kafka. There is a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support)
+for negative timestamps, but until that is implemented, KSQL can only support positive timestamps.
+* DATE and TIME types - because TIMESTAMPs represent a point in time, DATE and TIME types
+would be useful if a user wants to represent a less specific time. These would be added after
+TIMESTAMP is implemented though.
+* TIMESTAMP types that store timezone information (TIMESTAMP_TZ) - Several other databases support
+this. This is something that's useful to add in the future. 
+
+## Public APIS
+
+The TIMESTAMP data type will store a point in time after Unix epoch without timezone information.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (time TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+TIMESTAMPS will be displayed in console as strings in ODBC canonical format with millisecond precision:
+
+```roomsql
+> SELECT time FROM stream_name EMIT CHANGES;
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05 13:15:30:112 |
+```
+
+TIMESTAMPS can be represented by date strings:
+
+```roomsql
+INSERT INTO stream_name VALUES ("1994-11-05 13:15:30");
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the `java.sql.Timestamp` class in UTC within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it
+is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### Protobuf
+Protobuf 3 has a Timestamp type. The Protouf deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be deprecated:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
+
+These functions will still be available so that existing queries using these functions won't break,
+but the documentation will state that they are deprecated and will direct users towards using the
+TIMESTAMP type and the new functions.
+
+The following functions will be added:
+* `PARSE_TIMESTAMP(format, timestamp_string)` - converts a string into a TIMESTAMP 
+* `FORMAT_TIMESTAMP(format, timestamp)` - returns a string in the specified format
+* `NOW()` - returns the time after the issuing query is done executing
```

Comment:
```suggestion
* `NOW()` - returns the timestamp after the issuing query is done executing. All calls of now() within the same query return the same value.
```

Btw, I'm not sure about getting the time after the query is done executing. What if a user wants to use `NOW()` in the where clause to get the current day? 
I.e.
```
select * from stream where completed_date = format_timestamp(now(), 'YYYY-mm-dd');
```

Perhaps the now() needs to be obtained after the query is evaluated or planned? During the logical or physical plan? What if we use a push query (execution does not end until manually terminate it)?

Reply:
Updated to clarify this. Obtaining the `NOW` value seems to make the most sense in the logical plan.